File: legacy-reference/python/palermo_streets/entities/enemy.py
from __future__ import annotations

from dataclasses import replace
import random
import pygame

from .actor import Actor, AttackData
from ..config import PLAYFIELD
from ..systems.animation import AnimationBank


# Attacchi base. Ogni Enemy crea copie scalate dal profilo personaggio:
# in questo modo la Character Factory può registrare archetipi rapidi, pesanti
# o tecnici senza duplicare la logica dello stage.
ENEMY_ATTACK = AttackData("attack", 10, 0.48, 0.12, 0.56, 84, 56, 175, 0, 0, 0.050, 3.0)
ENEMY_HEAVY = AttackData("heavy", 16, 0.70, 0.15, 0.80, 100, 64, 250, 0, 0, 0.070, 6.0, knockdown=True)


def _scaled_attack(base: AttackData, damage_scale: float, speed_scale: float) -> AttackData:
    speed = max(0.55, min(1.80, speed_scale))
    damage = max(1, round(base.damage * max(0.35, damage_scale)))
    return replace(
        base,
        damage=damage,
        startup=base.startup / speed,
        active=base.active / speed,
        recovery=base.recovery / speed,
    )


class Enemy(Actor):
    def __init__(
        self,
        animations: AnimationBank,
        position: pygame.Vector2,
        health: int = 82,
        *,
        aggression: float = 1.0,
        boss: bool = False,
        display_name: str = "SGHERRO",
        variant_index: int = 0,
        character_id: str = "barbetta",
        visual_height: int = 275,
        move_speed_scale: float = 1.0,
        damage_scale: float = 1.0,
        attack_speed_scale: float = 1.0,
        heavy_chance: float = 0.13,
        cooldown_scale: float = 1.0,
        collision_scale: float = 1.0,
    ) -> None:
        super().__init__(animations, position, health, visual_height=visual_height)
        self.character_id = character_id
        self.aggression = aggression
        self.is_boss = boss
        self.display_name = display_name
        self.variant_index = variant_index
        self.move_speed_scale = max(0.60, min(1.65, move_speed_scale))
        self.cooldown_scale = max(0.55, min(1.80, cooldown_scale))
        self.collision_scale = max(0.75, min(1.35, collision_scale))
        self.heavy_chance = max(0.0, min(0.80, heavy_chance + (0.11 if boss else 0.0)))
        self.light_attack = _scaled_attack(ENEMY_ATTACK, damage_scale, attack_speed_scale)
        self.heavy_attack = _scaled_attack(ENEMY_HEAVY, damage_scale, attack_speed_scale)
        self.current_attack: AttackData | None = None
        self.attack_elapsed = 0.0
        self.attack_hit_player = False
        self.attack_cooldown = random.uniform(0.82, 1.22) * self.cooldown_scale
        self.attack_facing = -1
        self.preferred_depth_offset = random.uniform(-28.0, 28.0)
        self.engage_side: int | None = None
        self.spawn_elapsed = 0.0
        self.alpha = 0
        self.begin_state("spawn", "idle")

    @property
    def collision_radius(self) -> pygame.Vector2:
        if self.state in ("knockdown", "dead"):
            base = pygame.Vector2(62.0 if self.is_boss else 60.0, 16.0)
        elif self.state == "getup":
            base = pygame.Vector2(42.0 if self.is_boss else 40.0, 20.0)
        else:
            base = pygame.Vector2(43.0 if self.is_boss else 39.0, 23.0 if self.is_boss else 22.0)
        return base * self.collision_scale

    @property
    def attack_warning_ratio(self) -> float:
        attack = self.current_attack
        if self.state != "attack" or attack is None or attack.startup <= 0.0:
            return 0.0
        if self.attack_elapsed >= attack.startup:
            return 0.0
        return max(0.0, min(1.0, self.attack_elapsed / attack.startup))

    @property
    def attack_active(self) -> bool:
        attack = self.current_attack
        return bool(
            self.state == "attack"
            and attack is not None
            and attack.startup <= self.attack_elapsed <= attack.startup + attack.active
        )

    def _start_attack(self, attack: AttackData, player: Actor) -> None:
        self.current_attack = attack
        self.attack_elapsed = 0.0
        self.attack_hit_player = False
        self.attack_facing = 1 if player.position.x >= self.position.x else -1
        self.facing = self.attack_facing
        self.velocity.update(0, 0)
        self.begin_state("attack", attack.name)

    def update(
        self,
        dt: float,
        player: Actor,
        allies: list["Enemy"],
        *,
        may_attack: bool,
        support_rank: int = 0,
    ) -> None:
        self.update_common(dt)
        self.attack_cooldown = max(0.0, self.attack_cooldown - dt)

        if self.dead:
            # Lascia completare davvero la caduta prima della dissolvenza.
            # Le vecchie soglie fisse iniziavano il fade mentre alcuni archetipi
            # erano ancora a mezz'aria, rendendo le morti leggere e spezzate.
            dead_clip = self.animator.bank.clip("dead")
            landing_time = sum(frame.duration for frame in dead_clip.frames[:-1])
            hold = landing_time + (0.62 if self.is_boss else 0.34)
            if self.state_elapsed > hold:
                self.alpha = max(0, self.alpha - round((235 if self.is_boss else 430) * dt))
            if self.alpha <= 0:
                self.remove_ready = True
            return

        if self.state == "spawn":
            self.spawn_elapsed += dt
            duration = 0.44 if self.is_boss else 0.30
            self.alpha = min(255, round(255 * self.spawn_elapsed / duration))
            if self.spawn_elapsed >= duration:
                self.begin_state("idle", "idle")
                self.invulnerable = 0.15
            return

        if self.state == "hit":
            # Ogni personaggio può avere una reazione diversa: attendere la
            # fine della clip evita tagli dell'ultimo frame e salti di scala.
            if self.animator.finished:
                self.begin_state("idle", "idle")
            self._clamp()
            return

        if self.state == "knockdown":
            if self.animator.finished:
                self.begin_state("getup", "getup")
            self._clamp()
            return

        if self.state == "getup":
            if self.animator.finished:
                self.begin_state("idle", "idle")
                self.invulnerable = 0.14
            self._clamp()
            return

        if self.state == "attack" and self.current_attack is not None:
            attack = self.current_attack
            self.attack_elapsed += dt
            self.facing = self.attack_facing
            if attack.startup <= self.attack_elapsed <= attack.startup + attack.active:
                speed = 68.0 if attack.name == "heavy" else 48.0
                if self.is_boss:
                    speed *= 1.12
                self.position.x += self.attack_facing * speed * self.move_speed_scale * dt
            if self.attack_elapsed >= attack.total:
                self.current_attack = None
                base_low, base_high = ((0.66, 0.98) if self.is_boss else (0.96, 1.36))
                self.attack_cooldown = (
                    random.uniform(base_low, base_high)
                    * self.cooldown_scale
                    / max(0.78, self.aggression)
                )
                self.begin_state("idle", "idle")
            self._clamp()
            return

        delta = player.position - self.position
        if self.engage_side is None:
            self.engage_side = -1 if self.position.x < player.position.x else 1
        elif abs(delta.x) > 205.0:
            self.engage_side = -1 if self.position.x < player.position.x else 1

        if may_attack:
            side = self.engage_side
            desired_x = player.position.x + side * (96.0 if self.is_boss else 91.0)
            desired_y = player.position.y + self.preferred_depth_offset * 0.28
            self.facing = -side
        else:
            side = -1 if support_rank % 2 == 0 else 1
            ring = support_rank // 2
            distance = 176.0 + ring * 62.0
            lane_sign = -1 if ring % 2 == 0 else 1
            desired_x = player.position.x + side * distance
            desired_y = player.position.y + lane_sign * (50.0 + ring * 12.0)
            self.facing = 1 if player.position.x >= self.position.x else -1

        to_slot = pygame.Vector2(desired_x - self.position.x, desired_y - self.position.y)
        distance_x = abs(delta.x)
        distance_y = abs(delta.y)
        correct_side = (self.position.x - player.position.x) * self.engage_side > 0.0

        player_vulnerable = not player.dead and player.state not in ("hit", "knockdown", "getup")
        if (
            may_attack
            and player_vulnerable
            and correct_side
            and self.attack_cooldown <= 0.0
            and distance_x <= (118.0 if self.is_boss else 108.0)
            and distance_y <= 46.0
        ):
            chance = max(0.0, min(0.85, self.heavy_chance * self.aggression))
            attack = self.heavy_attack if random.random() < chance else self.light_attack
            self._start_attack(attack, player)
            self._clamp()
            return

        moved = False
        if to_slot.length_squared() > 12.0 ** 2:
            direction = to_slot.normalize()
            speed_x = ((132.0 if self.is_boss else 120.0) if may_attack else 92.0) * self.move_speed_scale
            speed_y = ((100.0 if self.is_boss else 92.0) if may_attack else 78.0) * self.move_speed_scale
            step = pygame.Vector2(direction.x * speed_x * dt, direction.y * speed_y * dt)
            if abs(step.x) > abs(to_slot.x):
                step.x = to_slot.x
            if abs(step.y) > abs(to_slot.y):
                step.y = to_slot.y
            self.position += step
            moved = step.length_squared() > 0.01

        for ally in allies:
            if ally is self or ally.dead:
                continue
            offset = self.position - ally.position
            if abs(offset.y) < 40.0 and abs(offset.x) < 90.0:
                sign = 1.0 if offset.x >= 0 else -1.0
                if abs(offset.x) < 1.0:
                    sign = 1.0 if self.actor_id > ally.actor_id else -1.0
                self.position.x += sign * (90.0 - abs(offset.x)) * 0.20

        self.animator.play("walk" if moved else "idle")
        self.state = "walk" if moved else "idle"
        self._clamp()

    def active_attack_box(self) -> pygame.Rect | None:
        attack = self.current_attack
        if attack is None or self.state != "attack" or self.attack_hit_player:
            return None
        if not (attack.startup <= self.attack_elapsed <= attack.startup + attack.active):
            return None
        center_x = self.position.x + self.attack_facing * attack.range_x * 0.50
        return pygame.Rect(
            round(center_x - attack.range_x / 2),
            round(self.position.y - attack.range_y),
            attack.range_x,
            attack.range_y,
        )

    def _clamp(self) -> None:
        self.clamp_to_playfield(PLAYFIELD)

File: src/game/entities/Enemy.ts
import { Actor, type HitResult } from './Actor';
import type { AnimationBank, AttackData, EnemyAttackPattern, Rect, Vec2 } from '../types';
import { ENEMY_ATTACK, ENEMY_HEAVY, attackTotal } from '../combat/attacks';
import { clamp, lengthSq, normalize, randomRange, sub } from '../../utils/math';
import { locomotionPlaybackRate, selectLocomotionClip } from '../animation/locomotion';
import { selectEnemyAttackSlot } from '../combat/enemyAttackPattern';

function scaledAttack(base: AttackData, damageScale: number, speedScale: number): AttackData {
  const speed = clamp(speedScale, 0.55, 1.8);
  return {
    ...base,
    damage: Math.max(1, Math.round(base.damage * Math.max(0.35, damageScale))),
    startup: base.startup / speed,
    active: base.active / speed,
    recovery: base.recovery / speed,
  };
}

export interface EnemyOptions {
  health?: number;
  aggression?: number;
  boss?: boolean;
  displayName?: string;
  variantIndex?: number;
  characterId?: string;
  moveSpeedScale?: number;
  damageScale?: number;
  attackSpeedScale?: number;
  heavyChance?: number;
  attackPattern?: EnemyAttackPattern;
  cooldownScale?: number;
  collisionScale?: number;
  dodgeChance?: number;
  dodgeCooldown?: number;
}

export class Enemy extends Actor {
  readonly characterId: string;
  readonly isBoss: boolean;
  readonly displayName: string;
  readonly variantIndex: number;
  readonly aggression: number;
  readonly moveSpeedScale: number;
  readonly cooldownScale: number;
  readonly collisionScale: number;
  readonly heavyChance: number;
  readonly attackPattern: EnemyAttackPattern;
  readonly lightAttack: AttackData;
  readonly heavyAttack: AttackData;
  readonly dodgeChance: number;
  readonly dodgeCooldownDuration: number;
  currentAttack: AttackData | null = null;
  attackElapsed = 0;
  attackHitPlayer = false;
  attackCooldown: number;
  attackFacing: -1 | 1 = -1;
  preferredDepthOffset = randomRange(-28, 28);
  engageSide: -1 | 1 | null = null;
  spawnElapsed = 0;
  grabbedBy: Actor | null = null;
  dodgeCooldown = 0;
  private attackSequence = 0;

  constructor(bank: AnimationBank, position: Vec2, options: EnemyOptions = {}) {
    super(bank, position, options.health ?? 82);
    this.characterId = options.characterId ?? 'talebano';
    this.aggression = options.aggression ?? 1;
    this.isBoss = options.boss ?? false;
    this.displayName = options.displayName ?? 'SGHERRO';
    this.variantIndex = options.variantIndex ?? 0;
    this.moveSpeedScale = clamp(options.moveSpeedScale ?? 1, 0.60, 1.65);
    this.cooldownScale = clamp(options.cooldownScale ?? 1, 0.55, 1.80);
    this.collisionScale = clamp(options.collisionScale ?? 1, 0.75, 1.35);
    this.heavyChance = clamp((options.heavyChance ?? 0.13) + (this.isBoss ? 0.11 : 0), 0, 0.8);
    this.attackPattern = options.attackPattern ?? 'weighted';
    this.lightAttack = scaledAttack(ENEMY_ATTACK, options.damageScale ?? 1, options.attackSpeedScale ?? 1);
    this.heavyAttack = scaledAttack(ENEMY_HEAVY, options.damageScale ?? 1, options.attackSpeedScale ?? 1);
    this.dodgeChance = clamp(options.dodgeChance ?? 0, 0, 0.8);
    this.dodgeCooldownDuration = clamp(options.dodgeCooldown ?? 2.6, 0.4, 12);
    this.attackCooldown = randomRange(0.82, 1.22) * this.cooldownScale;
    this.alpha255 = 0;
    this.beginState('spawn', 'idle');
  }

  override receiveHit(damage: number, knockback: Vec2, knockdown = false, launchVelocity = 0): HitResult {
    // A successful evade intentionally rejects the hit; combat resolution then
    // produces neither damage nor a false hit effect.
    const canDodge = this.animator.bank.clips.has('dodge')
      && this.canBeHit
      && this.dodgeCooldown <= 0
      && this.dodgeChance > 0
      && Math.random() < this.dodgeChance;
    if (!canDodge) return super.receiveHit(damage, knockback, knockdown, launchVelocity);
    const away = knockback.x === 0 ? -this.facing : (Math.sign(knockback.x) as -1 | 1);
    this.currentAttack = null;
    this.attackHitPlayer = false;
    this.velocity = { x: away * 245, y: 0 };
    this.facing = away === 1 ? -1 : 1;
    this.beginState('dodge', 'dodge');
    this.invulnerable = Math.max(0.30, this.animator.duration);
    this.dodgeCooldown = this.dodgeCooldownDuration;
    return { accepted: false, killed: false, knockedDown: false };
  }

  override get collisionRadius(): Vec2 {
    let base: Vec2;
    if (this.state === 'knockdown' || this.state === 'dead') base = { x: this.isBoss ? 62 : 60, y: 16 };
    else if (this.state === 'getup') base = { x: this.isBoss ? 42 : 40, y: 20 };
    else base = { x: this.isBoss ? 43 : 39, y: this.isBoss ? 23 : 22 };
    return { x: base.x * this.collisionScale, y: base.y * this.collisionScale };
  }

  get attackWarningRatio(): number {
    const attack = this.currentAttack;
    if (this.state !== 'attack' || !attack || attack.startup <= 0 || this.attackElapsed >= attack.startup) return 0;
    return clamp(this.attackElapsed / attack.startup, 0, 1);
  }

  get canBeGrabbed(): boolean {
    return !this.dead && this.invulnerable <= 0 && ['idle', 'walk'].includes(this.state);
  }

  beginGrabbed(holder: Actor): boolean {
    if (!this.canBeGrabbed) return false;
    this.grabbedBy = holder;
    this.velocity = { x: 0, y: 0 };
    this.currentAttack = null;
    this.beginState('grabbed', 'hit');
    this.animator.setPlaybackRate(0);
    return true;
  }

  releaseGrab(): void {
    if (!this.grabbedBy) return;
    this.grabbedBy = null;
    if (!this.dead && this.state === 'grabbed') this.beginState('idle', 'idle');
  }

  receiveGrabHit(damage: number): HitResult {
    if (this.dead || !this.grabbedBy) return { accepted: false, killed: false, knockedDown: false };
    this.health = Math.max(0, this.health - damage);
    this.hitFlash = 0.09;
    const killed = this.health <= 0;
    if (killed) {
      this.grabbedBy = null;
      this.dead = true;
      this.beginState('dead', 'dead');
      this.invulnerable = 999;
      return { accepted: true, killed: true, knockedDown: true };
    }
    this.beginState('grabbed', 'hit');
    this.animator.setPlaybackRate(0);
    return { accepted: true, killed: false, knockedDown: false };
  }

  private startAttack(attack: AttackData, player: Actor): void {
    this.currentAttack = attack;
    this.attackElapsed = 0;
    this.attackHitPlayer = false;
    this.attackFacing = player.position.x >= this.position.x ? 1 : -1;
    this.facing = this.attackFacing;
    this.velocity = { x: 0, y: 0 };
    this.beginState('attack', attack.name);
    this.animator.fitDuration(attackTotal(attack));
  }

  update(dt: number, player: Actor, allies: Enemy[], mayAttack: boolean, supportRank = 0): void {
    this.updateCommon(dt);
    this.attackCooldown = Math.max(0, this.attackCooldown - dt);
    this.dodgeCooldown = Math.max(0, this.dodgeCooldown - dt);

    if (this.grabbedBy) {
      if (this.grabbedBy.dead || !['grab', 'attack'].includes(this.grabbedBy.state)) {
        this.releaseGrab();
      } else {
        const holder = this.grabbedBy;
        this.position.x = holder.position.x + holder.facing * 54;
        this.position.y = holder.position.y;
        this.facing = holder.facing === 1 ? -1 : 1;
        this.velocity = { x: 0, y: 0 };
        this.state = 'grabbed';
        this.animator.setPlaybackRate(0);
        this.syncVisual();
        return;
      }
    }

    if (this.dead) {
      const deadClip = this.animator.bank.clips.get('dead')!;
      const landingTime = deadClip.frames.slice(0, -1).reduce((sum, frame) => sum + frame.duration, 0);
      const hold = landingTime + (this.isBoss ? 0.62 : 0.34);
      if (this.stateElapsed > hold) this.alpha255 = Math.max(0, this.alpha255 - (this.isBoss ? 235 : 430) * dt);
      if (this.alpha255 <= 0) this.removeReady = true;
      this.syncVisual();
      return;
    }

    if (this.state === 'spawn') {
      this.spawnElapsed += dt;
      const duration = this.isBoss ? 0.44 : 0.30;
      this.alpha255 = Math.min(255, Math.round(255 * this.spawnElapsed / duration));
      if (this.spawnElapsed >= duration) {
        this.beginState('idle', 'idle');
        this.invulnerable = 0.15;
      }
      this.syncVisual();
      return;
    }
    if (this.state === 'hit') {
      if (this.animator.finished) this.beginState('idle', 'idle');
      this.clampToPlayfield(); this.syncVisual(); return;
    }
    if (this.state === 'dodge') {
      if (this.animator.finished) this.beginState('idle', 'idle');
      this.clampToPlayfield(); this.syncVisual(); return;
    }
    if (this.state === 'knockdown') {
      if (this.animator.finished && this.elevation <= 0) this.beginState('getup', 'getup');
      this.clampToPlayfield(); this.syncVisual(); return;
    }
    if (this.state === 'getup') {
      if (this.animator.finished) {
        this.beginState('idle', 'idle');
        this.invulnerable = 0.14;
      }
      this.clampToPlayfield(); this.syncVisual(); return;
    }
    if (this.state === 'attack' && this.currentAttack) {
      const attack = this.currentAttack;
      this.attackElapsed += dt;
      this.facing = this.attackFacing;
      if (attack.startup <= this.attackElapsed && this.attackElapsed <= attack.startup + attack.active) {
        let speed = attack.name === 'heavy' ? 68 : 48;
        if (this.isBoss) speed *= 1.12;
        this.position.x += this.attackFacing * speed * this.moveSpeedScale * dt;
      }
      if (this.attackElapsed >= attackTotal(attack)) {
        this.currentAttack = null;
        const [low, high] = this.isBoss ? [0.66, 0.98] : [0.96, 1.36];
        this.attackCooldown = randomRange(low, high) * this.cooldownScale / Math.max(0.78, this.aggression);
        this.beginState('idle', 'idle');
      }
      this.clampToPlayfield(); this.syncVisual(); return;
    }

    const delta = sub(player.position, this.position);
    if (this.engageSide === null) this.engageSide = this.position.x < player.position.x ? -1 : 1;
    else if (Math.abs(delta.x) > 205) this.engageSide = this.position.x < player.position.x ? -1 : 1;

    let desiredX: number;
    let desiredY: number;
    if (mayAttack) {
      const side = this.engageSide;
      desiredX = player.position.x + side * (this.isBoss ? 96 : 91);
      desiredY = player.position.y + this.preferredDepthOffset * 0.28;
      this.facing = (-side) as -1 | 1;
    } else {
      const side = supportRank % 2 === 0 ? -1 : 1;
      const ring = Math.floor(supportRank / 2);
      const distance = 176 + ring * 62;
      const laneSign = ring % 2 === 0 ? -1 : 1;
      desiredX = player.position.x + side * distance;
      desiredY = player.position.y + laneSign * (50 + ring * 12);
      this.facing = player.position.x >= this.position.x ? 1 : -1;
    }

    const toSlot = { x: desiredX - this.position.x, y: desiredY - this.position.y };
    const distanceX = Math.abs(delta.x);
    const distanceY = Math.abs(delta.y);
    const correctSide = (this.position.x - player.position.x) * this.engageSide > 0;
    const playerVulnerable = !player.dead && !['hit', 'knockdown', 'getup'].includes(player.state);

    if (mayAttack && playerVulnerable && correctSide && this.attackCooldown <= 0 && distanceX <= (this.isBoss ? 118 : 108) && distanceY <= 46) {
      const chance = clamp(this.heavyChance * this.aggression, 0, 0.85);
      const selection = selectEnemyAttackSlot(this.attackPattern, this.attackSequence, chance);
      this.attackSequence = selection.nextSequence;
      this.startAttack(selection.slot === 'heavy' ? this.heavyAttack : this.lightAttack, player);
      this.clampToPlayfield(); this.syncVisual(); return;
    }

    let moved = false;
    let movementAnimation = 'walk';
    if (lengthSq(toSlot) > 12 * 12) {
      const direction = normalize(toSlot);
      const speedX = (mayAttack ? (this.isBoss ? 132 : 120) : 92) * this.moveSpeedScale;
      const speedY = (mayAttack ? (this.isBoss ? 100 : 92) : 78) * this.moveSpeedScale;
      const step = { x: direction.x * speedX * dt, y: direction.y * speedY * dt };
      if (Math.abs(step.x) > Math.abs(toSlot.x)) step.x = toSlot.x;
      if (Math.abs(step.y) > Math.abs(toSlot.y)) step.y = toSlot.y;
      this.position.x += step.x;
      this.position.y += step.y;
      moved = lengthSq(step) > 0.01;
      if (moved) {
        movementAnimation = selectLocomotionClip(
          step,
          this.animator.name,
          (name) => this.animator.bank.clips.has(name),
        );
        const actualSpeed = Math.hypot(direction.x * speedX, direction.y * speedY);
        const clip = this.animator.bank.clips.get(movementAnimation);
        this.animator.setPlaybackRate(locomotionPlaybackRate(actualSpeed, clip?.referenceSpeed));
      }
    }

    for (const ally of allies) {
      if (ally === this || ally.dead) continue;
      const offset = sub(this.position, ally.position);
      if (Math.abs(offset.y) < 40 && Math.abs(offset.x) < 90) {
        let sign = offset.x >= 0 ? 1 : -1;
        if (Math.abs(offset.x) < 1) sign = this.actorId > ally.actorId ? 1 : -1;
        this.position.x += sign * (90 - Math.abs(offset.x)) * 0.20;
      }
    }

    const preservesStride = ['walk', 'walk_up', 'walk_down'].includes(this.animator.name);
    this.animator.play(moved ? movementAnimation : 'idle', false, moved && preservesStride);
    if (!moved) this.animator.setPlaybackRate(1);
    this.state = moved ? 'walk' : 'idle';
    this.clampToPlayfield();
    this.syncVisual();
  }

  activeAttackBox(): Rect | null {
    const attack = this.currentAttack;
    if (!attack || this.state !== 'attack' || this.attackHitPlayer) return null;
    if (!(attack.startup <= this.attackElapsed && this.attackElapsed <= attack.startup + attack.active)) return null;
    const centerX = this.position.x + this.attackFacing * attack.rangeX * 0.50;
    return { x: centerX - attack.rangeX / 2, y: this.position.y - attack.rangeY, width: attack.rangeX, height: attack.rangeY };
  }
}
